Taarafo.Core/Services/Processings/PostImpressions/PostImpressionProcessingService.cs

```diff
@@ -23,6 +26,31 @@ public partial class PostImpressionProcessingService : IPostImpressionProcessing
             this.loggingBroker = loggingBroker;
         }
 
+        public ValueTask<PostImpression> UpsertPostImpressionAsync(PostImpression postImpression) =>
+        TryCatch(async () =>
+        {
+            ValidatePostImpression(postImpression);
+            PostImpression maybePostImpression = RetrieveMatchingPostImpression(postImpression);
+
+            return maybePostImpression switch
+            {
+                null => await this.postImpressionService.AddPostImpressions(postImpression),
+                _ => await this.postImpressionService.ModifyPostImpressionAsync(postImpression)
+            };
+        });
+
+        private PostImpression RetrieveMatchingPostImpression(PostImpression postImpression)
+        {
+            IQueryable<PostImpression> postImpressions =
+                this.postImpressionService.RetrieveAllPostImpressions();
+
+            return postImpressions.FirstOrDefault(SamePostImpressionAs(postImpression));
+        }
+
+        private static Expression<Func<PostImpression, bool>> SamePostImpressionAs(PostImpression postImpression) =>
+            retrievePostImpression => (retrievePostImpression.PostId == postImpression.PostId)
+                || (retrievePostImpression.ProfileId == postImpression.ProfileId);
```

Comment:
You need to macth on both.  Change `||` to `&&`

Reply:
Done